warehouse/query-core/src/main/java/datawave/query/jexl/DatawavePartialInterpreter.java

```diff
@@ -0,0 +1,887 @@
+        State leftState = null;
+        State rightState = null;
+        
+        if (left != null) {
+            if (left instanceof State) {
+                leftState = (State) left;
+            } else if (left instanceof FunctionalSet) {
+                leftState = new State(true, false, (FunctionalSet) left);
+            }
+        }
+        
+        if (right != null) {
+            if (right instanceof State) {
+                rightState = (State) right;
+            } else if (right instanceof FunctionalSet) {
+                rightState = new State(true, false, (FunctionalSet) right);
+            }
+        }
+        
+        if (leftState != null && rightState != null) {
+            leftState.mergeOr(rightState);
+            return leftState;
+        } else if (leftState != null) {
+            return leftState;
+        } else if (rightState != null) {
+            return rightState;
+        } else {
+            // if we got here then we are likely dealing with a function that takes a union of null fields as arguments
+            // for example, filter:notNull(NULL1 || NULL2). return false with an initialized empty functional set
+            return new State(true);
+        }
+    }
+    
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
+        // we could have arrived here after the node was dereferenced
+        if (QueryPropertyMarker.findInstance(node).isType(ExceededOrThresholdMarkerJexlNode.class)) {
+            Object o = visitExceededOrThresholdMarker(node);
+            return getState(node, o);
+        }
+        
+        // check for the special case of a range (conjunction of a G/GE and a L/LE node) and reinterpret as a function
+        Object evaluation = evaluateRange(node);
+        if (evaluation != null) {
+            return getState(node, evaluation);
+        }
+        
+        // values for intersection
+        boolean anyFromUnknown = false;
+        FunctionalSet functionalSet = new FunctionalSet();
+        for (JexlNode child : JexlNodes.children(node)) {
+            
+            Object o = child.jjtAccept(this, data);
+            if (o instanceof State) {
+                State state = (State) o;
+                if (!state.matched) {
+                    return state;
+                } else {
+                    if (state.fromUnknown) {
+                        anyFromUnknown = true;
+                    }
+                    if (state.set != null && !state.set.isEmpty()) {
+                        functionalSet.addAll(state.set);
+                    }
+                }
+            } else {
+                throw new IllegalStateException("expected STATE but was " + o.getClass().getSimpleName());
+            }
+        }
+        
+        if (!functionalSet.isEmpty()) {
+            return new State(true, anyFromUnknown, functionalSet);
+        } else {
+            return new State(true, anyFromUnknown);
+        }
+    }
```

Comment:
should these be `new State(o.matched, ...)`?

Reply:
No. At this point in the method the intersection has evaluated to true. The only remaining questions are a) is there a functional set of hits and b) was any part of this from an unknown/incomplete field